Change votes to enum

Murph33 opened the pull request:
Change the voting model to go from taking in a value and putting that into the database to taking in a string or symbol and using a hash located on the model to convert that to a value for the database.  I18n is utilized as well instead of a helper method as was the previous implementation.

Murph33 committed:
Change voting model to use an enum for values.

Remove the constant VALID_VALUES, VOTE_TYPES and the custome validation
from the vote model as they are no longer needed.  Tests were changed to
the reflect the fact an ArgumentError is raised when an invalid
parameter is passed to create or update.  The test for the quote
object's score function was also modified to use the enum.  I18n was
utilized to translate the enum key into a more user friendly sentence.

Murph33 committed:
Clean votes controller to reflect enum change

The update action was refactored to remove the flash message because
errors will be thrown if either the vote was already deleted or else an
invalid parameter is passed.

Murph33 committed:
Change hardcoded test to utilize variables

This will keep us from having to change the test later by hand in case
we change the values related to the hash or expand the hash for voting.

seantaylor commented on app/controllers/votes_controller.rb:
```diff
@@ -3,22 +3,15 @@ class VotesController < ApplicationController
-    end
-    redirect_to quotes_path
+    vote.save
+    redirect_to quotes_path, notice: "Voted!" # error thrown if save fails
```
`.save` won't actually throw an error if it fails. You're looking for `.save!`. Although I prefer to `if/else` on the `.save` and show validation errors to the user in most cases.

Murph33 replied:
Ah yeah I was trying to be too clever for my own good.  It will throw the error if the value is outside of the accepted but if someone has a couple tabs open and votes in one and moves to the other and votes again it will keep the first score and not alert them that their new vote didn't count.  Will change this UX around.

Murph33 committed:
Add additional tests for vote creation

Added tests for condition where a vote already exists in the database
for the user for that quote.

jarednorman commented on config/locales/en.yml:
```diff
@@ -21,3 +21,9 @@
 
 en:
   hello: "Hello world"
+  votes: {
```
Don't need these curlies.

Murph33 replied:
It breaks without them.  Maybe since I'm calling .each on t(:votes) ?

jarednorman replied:
Still waiting for the curlies to disappear.

Murph33 replied:
I tried taking them out and it throws an error.  

At least I'm assuming you just mean to take them out and leave everything else the same cause that's what I tried to do.

Murph33 replied:
ah.  yaml doesn't like ,s apparently.  one of those whitespace things.  

thanks richard for the help

jarednorman replied:
I knew you were forgetting the commas, but thought it better to for you to figure it out yourself. 

@Senjai quit being so helpful.

jarednorman commented on app/views/votes/_voting.html.erb:
```diff
@@ -1,11 +1,10 @@
   <%= link_to "Erase my vote", quote_vote_path(quote, vote), method: :delete %>
 <% else %>
-  <% Vote::VOTE_TYPES.each do |reason, value| %>
+  <% t(:votes).each do |value, reason| %>
```
Storing objects in the locales isn't a conventional thing to do. Try a different approach.

Murph33 replied:
This wasn't my first idea (or something I was aware of) but it is what @Senjai had recommended as something that was somewhat standard.  Would you recommend a method or constant on the model or a helper method?  Something else entirely?

Murph33 replied:
I don't know if it pings you just from me referencing this PR somewhere else but I did this a different, possibly better way in #32 .  Didn't push over this since I didn't want to totally erase the I18n stuff and it was still possibly being discussed?  I do like the new method implemented though.

Senjai replied:
To be fair, I said you should translate the reason. Not store the votes. Vote::VOTE_TYPES.each do |reason, value|, and then translate the value or reason.

jarednorman replied:
Yeah, list should be elsewhere, probably as symbols, and just have the regular translations in the locale file.

Murph33 replied:
Ah.  Sorry about that @Senjai , misunderstood how you meant to implement it.

Murph33 replied:
Okay so I implemented this in a different way now..  The way Richard wanted me to.  Please check that out.  Probably better than what I did in #32 because while we do have to mess around in two files here it gives us more flexibility than the other implementation.

jarednorman commented on app/models/vote.rb:
```diff
@@ -1,21 +1,9 @@
+  enum value: { hate_it: -2, dislike_it: -1, like_it: 1, love_it: 2 }
 
-  validates :value, presence: true
+  validates :value, presence: true, inclusion: { in: Vote.values.keys }
```
Is this inclusion validation required? I would assume that `enum` does it out of the box, but I'm just making assumptions.

Murph33 replied:
You are correct.

jarednorman commented on app/views/votes/_voting.html.erb:
```diff
@@ -1,11 +1,10 @@
-  <% Vote::VOTE_TYPES.each do |reason, value| %>
-    <%= link_to reason, quote_vote_path(quote, vote, { vote: { value: value } } ), method: :patch %>
+  <% Vote.values.keys.each do |opinion| %>
+    <%= link_to t(:votes)[opinion.to_sym], quote_vote_path(quote, vote, { vote: { value: opinion } } ), method: :patch %>
```
use the `scope:` option for the `t` helper here.

jarednorman commented on app/views/votes/_voting.html.erb:
```diff
@@ -1,11 +1,10 @@
-  <% Vote::VOTE_TYPES.each do |reason, value| %>
-    <%= link_to reason, quote_votes_path(quote, { vote: { value: value } } ), method: :post %>
+  <% Vote.values.keys.each do |opinion| %>
+    <%= link_to t(:votes)[opinion.to_sym], quote_votes_path(quote, { vote: { value: opinion } } ), method: :post %>
```
and of course here

jarednorman commented on app/views/votes/_voting.html.erb:
```diff
@@ -1,11 +1,10 @@
-  <% Vote::VOTE_TYPES.each do |reason, value| %>
-    <%= link_to reason, quote_vote_path(quote, vote, { vote: { value: value } } ), method: :patch %>
+  <% Vote.values.keys.each do |opinion| %>
+    <%= link_to t(opinion, scope: [:votes]), quote_vote_path(quote, vote, { vote: { value: opinion } } ), method: :patch %>
```
pretty sure you can do just `t(opinion, scope: :votes)`

jarednorman replied:
also you've got a space inside a `)` which is not a thing

Murph33 replied:
right again will try to keep everything cleaner.  Thanks again.

Sinetheta commented:
Good stuff, Jared's feedback has been actioned, specs pass, app works.
